crates/noirc_abi/src/input_parser/toml.rs

```diff
@@ -1,17 +1,26 @@
 use super::InputValue;
-use crate::errors::InputParserError;
+use crate::{errors::InputParserError, Abi, AbiType};
 use acvm::FieldElement;
 use serde::Serialize;
 use serde_derive::Deserialize;
 use std::collections::BTreeMap;
 
 pub(crate) fn parse_toml(
     input_string: &str,
+    mut abi: Abi,
 ) -> Result<BTreeMap<String, InputValue>, InputParserError> {
     // Parse input.toml into a BTreeMap, converting the argument to field elements
     let data: BTreeMap<String, TomlTypes> = toml::from_str(input_string)
         .map_err(|err_msg| InputParserError::ParseTomlMap(err_msg.to_string()))?;
-    toml_map_to_field(data)
+
+    // The toml map is stored in an ordered BTreeMap. As the keys are strings the map is in alphanumerical order.
+    // When parsing the toml map we recursively go through each field to enable struct inputs.
+    // To match this map with the correct abi type we must sort our abi parameters and then flatten each struct.
+    abi.sort();
+    let flat_abi_types = abi.flattened_param_types();
```

Comment:
Passing a `BTreeMap<String, AbiType>` seems more natural here. We can get rid of the index counting and it'll be easier to reconcile with #677 (as then we can just pass the `AbiType`).

Reply:
Good note. I went with the `Vec<AbiType>` originally as the parameter name key in the toml map does not distinguish between structs and I made an assumption that this method would be simpler. However, reconstructing the Abi into a `BTreeMap<String, AbiType>` mocking the toml map as you said is definitely cleaner. I have changed this to use a `BTreeMap` now